tests/regression/src/lib.rs

```diff
@@ -105,52 +122,135 @@ mod tests {
+        command
+    }
 
-        println!("Running command: {:?}", command);
+    fn execute_command(mut command: Command) {
         let status = command.status().expect("Failed to execute valgrind");
-
         if !status.success() {
             panic!("Valgrind failed");
         }
+    }
 
+    fn run_cg_annotate(output_file: &str) -> std::process::Output {
         let annotate_output = Command::new("cg_annotate")
-            .arg(&output_file)
+            .arg(output_file)
             .output()
             .expect("Failed to run cg_annotate");
-
         if !annotate_output.status.success() {
             panic!("cg_annotate failed");
         }
-        create_dir_all(Path::new("target/perf_outputs")).unwrap();
-        let annotate_file = format!("target/perf_outputs/{}.annotated.txt", test_name);
-        let mut file = File::create(&annotate_file).expect("Failed to create annotation file");
-        file.write_all(&annotate_output.stdout)
+        annotate_output
+    }
+
+    /// Saves the annotated output to prev, curr, or diff accordingly
+    fn save_annotate_output(output: &std::process::Output, suffix: &str, test_name: &str) {
+        let directory = format!("target/perf_outputs/{}", suffix);
+        create_dir_all(Path::new(&directory)).unwrap();
+        let annotate_file = format!(
+            "target/perf_outputs/{}/{}_{}.annotated.txt",
+            suffix, test_name, suffix
+        );
+        let mut file = File::create(annotate_file).expect("Failed to create annotation file");
+        file.write_all(&output.stdout)
             .expect("Failed to write annotation file");
+    }
 
-        let count = find_instruction_count(&annotate_file)
-            .expect("Failed to get instruction count from file");
-        // This is temporary code to showcase the future diff functionality, here the code regresses by 10% each time so this test will almost always fail
-        let new_count = count + count / 10;
-        let diff = new_count - count;
-        assert!(diff <= self::MAX_DIFF, "Instruction count difference in {} exceeds the threshold, regression of {} instructions", test_name, diff);
+    /// Function to run the diff test using valgrind, only called when diff mode is set
+    fn run_diff_test(test_name: &str) {
+        let (prev_file, curr_file) = get_diff_files(test_name);
+        ensure_diff_files_exist(&prev_file, &curr_file);
+
+        let diff_output = run_cg_annotate_diff(&prev_file, &curr_file);
+        save_diff_output(&diff_output, test_name);
+
+        let diff = find_instruction_count(&diff_output)
+            .expect("Failed to parse cg_annotate --diff output");
+
+        assert_diff_within_threshold(diff, test_name);
+    }
+
+    fn get_diff_files(test_name: &str) -> (String, String) {
+        (
+            format!("target/cg_artifacts/cachegrind_{}_prev.out", test_name),
+            format!("target/cg_artifacts/cachegrind_{}_curr.out", test_name),
```

Comment:
```suggestion
            format!("target/cg_artifacts/cachegrind_{test_name}_prev.out"),
            format!("target/cg_artifacts/cachegrind_{test_name}_curr.out"),
```

Reply:
Instead of assuming `prev` and `curr`, can we pull these from the environment? I think it would ultimately be nice to use commit hashes as the names.

Reply:
Hmm yeah, that would be nice. I had avoided that because in CI I wasn't sure if we'd be able to reference the commit ID's but it seems Github has [environment variables](https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables) for that. I'll implement a solution, although I still think it will be necessary to store the build artifacts in a pre-defined path which differentiates one commit from the other so that the `get_diff_files` knows where to look without making the user identify the commit hashes. The final outputs files would still be named by the commit hash though which adds a greater level of insight and extensibility for the tooling.